### ISSUE TITLE
Documentation

### ISSUE THREAD

**Issue by fluidnumerics-joe** [ROCm/hipfort#41]
Hey all, I've started a branch (https://github.com/ROCmSoftwarePlatform/hipfort/tree/repo-docs) where we can start working together on building this repository's documentation.

* [CONTRIBUTING.md](https://github.com/ROCmSoftwarePlatform/hipfort/blob/repo-docs/CONTRIBUTING.md)
* [SPECIFICATIONS.md](https://github.com/ROCmSoftwarePlatform/hipfort/blob/repo-docs/SPECIFICATIONS.md)
* [readthedocs](https://github.com/ROCmSoftwarePlatform/hipfort/tree/repo-docs/docs)

It'd be great if we can be transparent in the specifications about the assumptions and expectations we have for end-users of hipfort and provide clear guidelines for revising hipfort specifications with the community.

To lower the barrier for community engagement for contributing to the hipfort repository, we should spend some time thinking about our branching model, versioning model, and process for moving branches from features and bug fixes and into develop and main branches. Documenting this and publishing to this repository will help users and potential contributors more easily see how we get work done on hipfort.

To start, here are some questions to spur discussion around contributing 
* What are our acceptance criteria for getting branches into develop, main, or even into tagged releases ? 
* How can we enable others to test their contributions before opening pull requests ?
* Do we want to automate testing ?
* What information is necessary when opening feature requests or bug reports ? Do we want to leverage issue templates ?

**Comment by txomin13** [comment]
@schoonovernumerics I added a doxygen target to the sphinx Makefile.  
Would be interested in using `breathe` to integrate this docu into the sphinx doc you added.
Do you have some experience with that?

![image](https://user-images.githubusercontent.com/19530127/107811761-f9d13680-6d6e-11eb-8098-0b3538ec33f9.png)

**Comment by schoonovernumerics** [comment]
I have not used breathe before but will look into it. Thanks for setting this up. This is a good step in the right direction. From what I recall with Doxygen, we can annotate source code in comments to pull in additional information. I'm thinking that for a number of the functions, having example usage would be helpful. I'll pull down what you've started here and see what we can do to accompany the doxygen API documentation.

**Comment by txomin13** [comment]
@schoonovernumerics , all modules are automatically generated from the C/C++ HIP and ROCm header files.
If there is any documentation in these files, it will be attached to the respective interface.

Interestingly, the roc* libaries, particularly rocsolver, provide more docu than the hip* ones.
Here is an example of rocsolver:

![image](https://user-images.githubusercontent.com/19530127/107818922-9a782400-6d78-11eb-9e43-1303dec3762b.png)

**Comment by txomin13** [comment]
Each interface has a number of additional routines that internally call the `_orig` or `_b` routine, which is directly bound to a C function and takes `type(c_ptr)`
arguments. These additional routines, typically called `_rank0`, `_rank1`, `full_rank`, directly take Fortran pointers
with rank 0 (scalars), rank 1 (vectors) or full rank (a mix of matrix and vector arguments where appropriate.  
We can certainly add some more docu to these routines that explains their respective suffix.

**Comment by txomin13** [comment]
@schoonovernumerics I am currently using Github Pages to host the doxygen docu associated with hipfort. 

You can find the docu under the following address:
https://rocmsoftwareplatform.github.io/hipfort

README.md hipfort webpage URL both refer to this page now.

**Comment by schoonovernumerics** [comment]
Is there a way to bring in additional documentation to the github.io page, in addition to the API documentation from doxygen ? Things like, how to's and basic tutorials/demos.

**Comment by txomin13** [comment]
Yes, it is possible. Github Pages just looks for a `README.md` or `index.html` at the specified location.
My plan would be to integrate the doxygen docu with the sphinx guide via breathe and
then point Github Pages to the sphinx guide's index.tml

**Comment by txomin13** [comment]
We have to work a little on the guide first before this makes sense.

**Comment by cgmb** [comment]
> Interestingly, the roc* libaries, particularly rocsolver, provide more docu than the hip* ones.

This was because the HIP library interfaces should be compatible with CUDA library interfaces and it seemed redundant with both the roc* library docs and CUDA library docs. With that said, I think that decision was probably a mistake. I believe that has been something we've been trying to improve (albeit slowly).

I'm a bit uncertain of what the hipfort documentation should look like. It seems silly that we duplicate all the library documentation in hipfort, particularly as the interfaces are being presented in the hipfort docs as if they were C interfaces. Ideally, the hipfort docs would probably be the opposite. We'd want to convey the Fortran-specific information for each function in the hipfort docs, and maybe just link to the actual library docs for the detailed description of the function semantics.

**Comment by cgmb** [comment]
I'm going to close this, just because the implementation of the documentation has changed so much that a lot of this discussion is no longer relevant. The documentation still needs improvements and there are several pieces of this issue that are still relevant, but I think we may want to start some new issues focused on particular aspects of the docs.

A CONTRIBUTING guide with acceptance criteria is certainly still useful. I believe that there's now a general set of contributing guidelines for the AMD ROCm libraries that we can add project-specific information to. There are also bug report templates for all ROCm libraries. It would also be good to update our README with the target compilers as you suggested.
